ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -543,27 +544,45 @@ fromList k n (anchor, signers, ebbs) =
   Serialization
 -------------------------------------------------------------------------------}
 
+serializationFormatVersion1 :: Word8
+serializationFormatVersion1 = 1
+  -- CHANGELOG
+  --
+  -- Version 0 is 2 fields, the anchor and the window. Note that it does not
+  -- have the version marker.
+  --
+  -- Version 1 has 4 fields, the version marker, anchor, window, and EbbMap.
+
 encodePBftChainState :: (PBftCrypto c, Serialise (PBftVerKeyHash c))
                      => PBftChainState c -> Encoding
 encodePBftChainState st@PBftChainState{..} = mconcat [
-      Serialise.encodeListLen 3
+      Serialise.encodeListLen 4
+    , encode serializationFormatVersion1
     , encode (withOriginToMaybe anchor)
     , encode signers
     , encode ebbs'
     ]
   where
     (anchor, signers, ebbs') = toList st
 
-decodePBftChainState :: (PBftCrypto c, Serialise (PBftVerKeyHash c))
+decodePBftChainState :: (PBftCrypto c, Serialise (PBftVerKeyHash c), HasCallStack)
                      => SecurityParam
                      -> WindowSize
                      -> Decoder s (PBftChainState c)
-decodePBftChainState k n = do
-      Serialise.decodeListLenOf 3
+decodePBftChainState k n = Serialise.decodeListLen >>= \case
+   2 -> do -- Version is 0
+      anchor  <- withOriginFromMaybe <$> decode
+      signers <- decode
+      return $ fromList k n (anchor, signers, ebbsEmpty)
+   4 -> do -- Version is >0
```

Comment:
Hmm, what if the next version (I don't think we'll need one, but hypothetically speaking) happens to need more or fewer fields, e.g., 5? Then the `listLen` is incorrect (I know, *I* suggested this approach :slightly_smiling_face: ). I'm thinking now that it would be better to put the version number first (make sure it has the same size as the `encodeListLen` tag) and reserve version 2 (= `encodeListLen 2`) for the old format. Let's call the new format version 3 (newer version -> higher number). In the new format, the version number is then followed by another, proper, `encodeListLen 3`. Or does that go against the rules of CBOR?

Thoughts?

Reply:
Thomas and I talked about it on Slack. We concluded:

  * "The list ship as sailed" insofar as we want to preserve the fact that the chain state's outermost structure as a CBOR term is that of a list.
  * As long as no future version of the format has exactly 2 fields, the fact that at least one field will always be an explicit version marker ensures future-proofness and backwards-compatibility are possible.
  * Byte size isn't a large concern, so there's significant cost to including a version marker.